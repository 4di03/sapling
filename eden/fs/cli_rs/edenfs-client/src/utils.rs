/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::env;
use std::path::Path;
use std::path::PathBuf;

use anyhow::anyhow;
use anyhow::bail;
use anyhow::Context;
use anyhow::Result;
use hg_util::path::expand_path;

use crate::instance::DEFAULT_CONFIG_DIR;
use crate::instance::DEFAULT_ETC_EDEN_DIR;

/// Traverse up and locate the repository root
pub fn locate_repo_root(path: &Path) -> Option<&Path> {
    path.ancestors()
        .find(|p| p.join(".hg").is_dir() || p.join(".git").is_dir())
}

pub fn get_mount_point(mount_point: &Option<PathBuf>) -> Result<PathBuf> {
    if let Some(path) = mount_point {
        Ok(path.clone())
    } else {
        locate_repo_root(
            &std::env::current_dir().context("Unable to retrieve current working directory")?,
        )
        .map(|p| p.to_path_buf())
        .ok_or_else(|| anyhow!("Unable to locate repository root"))
    }
}

/// Expand the path if the user has supplied anything. Otherwise, use the current working directory instead.
///
/// Usage:
/// ```ignore
/// #[clap(..., parse(try_from_str = expand_path_or_cwd), default_value = "", ...)]
/// ```
pub fn expand_path_or_cwd(input: &str) -> Result<PathBuf> {
    // TODO(T135653638): This function must be updated when expand_path is updated to use OsString
    if input.is_empty() {
        std::env::current_dir().context("Unable to retrieve current working directory")
    } else {
        Ok(expand_path(input))
    }
}

/// Utility function to remove trailing slashes from user provided relative paths. This is required
/// because some EdenFS internals do not allow trailing slashses on relative paths.
pub fn remove_trailing_slash(path: &Path) -> PathBuf {
    // TODO(T135653638): This function must be updated when expand_path is updated to use OsString
    PathBuf::from(
        path.to_string_lossy()
            .trim_end_matches(if cfg!(windows) { r"\" } else { "/" }),
    )
}

pub fn locate_eden_config_dir(path: &Path) -> Option<PathBuf> {
    // Check whether we're in an Eden mount. If we are, some parent directory will contain
    // a .eden dir that contains a socket file. This socket file is symlinked to the
    // socket file contained in the config dir we should use for this mount.
    if let Ok(expanded_path) = path.canonicalize() {
        for ancestor in expanded_path.ancestors() {
            let socket = ancestor.join(".eden").join("socket");
            if socket.exists() {
                if let Ok(resolved_socket) = socket.canonicalize() {
                    if let Some(parent) = resolved_socket.parent() {
                        return Some(parent.to_path_buf());
                    }
                }
            }
        }
    }
    None
}

pub fn get_config_dir(
    config_dir_override: &Option<PathBuf>,
    mount_path_override: &Option<PathBuf>,
) -> Result<PathBuf> {
    // A config dir might be provided as a top-level argument. Top-level arguments take
    // precedent over sub-command args.
    if let Some(config_dir) = config_dir_override {
        if config_dir.as_os_str().is_empty() {
            bail!("empty --config-dir path specified")
        }
        Ok(config_dir.clone())
    // Then check if the optional mount path provided by some subcommands is an EdenFS mount.
    // If it's provided and is a valid EdenFS mount, use the mounts config dir.
    } else if let Some(config_dir) = mount_path_override
        .as_ref()
        .and_then(|x| locate_eden_config_dir(x))
    {
        Ok(config_dir)
    // Then check if the current working directory is an EdenFS mount. If not, we should
    // default to the default config-dir location which varies by platform.
    } else {
        Ok(env::current_dir()
            .map_err(From::from)
            .and_then(|cwd| {
                locate_eden_config_dir(&cwd).ok_or_else(|| anyhow!("cwd is not in an eden mount"))
            })
            .unwrap_or(expand_path(DEFAULT_CONFIG_DIR)))
    }
}

pub fn get_etc_eden_dir(etc_eden_dir_override: &Option<PathBuf>) -> PathBuf {
    if let Some(etc_eden_dir) = etc_eden_dir_override {
        etc_eden_dir.clone()
    } else {
        DEFAULT_ETC_EDEN_DIR.into()
    }
}

pub fn get_home_dir(home_dir_override: &Option<PathBuf>) -> Option<PathBuf> {
    if let Some(home_dir) = home_dir_override {
        Some(home_dir.clone())
    } else {
        dirs::home_dir()
    }
}

/// Given a prefix and a list of paths, return a list of paths with the prefix prepended to each path.
///
/// If the prefix is None the paths are processed as-is.
/// All paths are post-processed with the provided function.
pub(crate) fn prefix_paths<F, T>(
    prefix: &Option<PathBuf>,
    paths: &Option<Vec<PathBuf>>,
    f: F,
) -> Option<Vec<T>>
where
    F: Fn(PathBuf) -> T,
{
    if let Some(prefix) = prefix {
        paths
            .as_ref()
            .map(|ps| ps.iter().map(|p| f(prefix.join(p))).collect::<Vec<_>>())
    } else {
        paths
            .as_ref()
            .map(|ps| ps.iter().map(|p| f(p.to_path_buf())).collect::<Vec<_>>())
    }
}

/// Given a prefix and a path string, return the path with the prefix removed.
///
/// If the prefix is None, the path is returned as-is.
pub fn strip_prefix_from_string(prefix: &Option<PathBuf>, path: String) -> String {
    if let Some(prefix) = prefix {
        let path = Path::new(&path);
        path.strip_prefix(prefix)
            .map_or(path, |stripped_path| stripped_path)
            .to_string_lossy()
            .to_string()
    } else {
        path
    }
}

pub(crate) fn strip_prefix_from_bytes(prefix: &Option<PathBuf>, path: &[u8]) -> Vec<u8> {
    if let Some(prefix) = prefix {
        let path = Path::new(std::str::from_utf8(path).expect("Failed to convert path to string"));
        path.strip_prefix(prefix)
            .map_or(path, |stripped_path| stripped_path)
            .to_string_lossy()
            .to_string()
            .into_bytes()
    } else {
        path.to_vec()
    }
}
