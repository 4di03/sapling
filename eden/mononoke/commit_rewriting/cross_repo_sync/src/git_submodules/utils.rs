/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::collections::HashMap;
use std::collections::HashSet;
use std::str::FromStr;

use anyhow::anyhow;
use anyhow::Context;
use anyhow::Error;
use anyhow::Result;
use context::CoreContext;
use fsnodes::RootFsnodeId;
use futures::future;
use futures::stream;
use futures::stream::Stream;
use futures::stream::StreamExt;
use futures::stream::TryStreamExt;
use futures::Future;
use futures_stats::TimedFutureExt;
use git_types::ObjectKind;
use manifest::bonsai_diff;
use manifest::BonsaiDiffFileChange;
use manifest::Entry;
use manifest::ManifestOps;
use mononoke_types::hash::GitSha1;
use mononoke_types::hash::RichGitSha1;
use mononoke_types::ChangesetId;
use mononoke_types::ContentId;
use mononoke_types::FileType;
use mononoke_types::FsnodeId;
use mononoke_types::MPathElement;
use mononoke_types::NonRootMPath;
use repo_blobstore::RepoBlobstoreArc;
use repo_derived_data::RepoDerivedDataRef;

use crate::git_submodules::expand::SubmodulePath;
use crate::git_submodules::in_memory_repo::InMemoryRepo;
use crate::types::Repo;

/// Get the git hash from a submodule file, which represents the commit from the
/// given submodule that the source repo depends on at that revision.
pub(crate) async fn get_git_hash_from_submodule_file<'a, R: Repo>(
    ctx: &'a CoreContext,
    repo: &'a R,
    submodule_file_content_id: ContentId,
) -> Result<GitSha1> {
    let blobstore = repo.repo_blobstore_arc();

    let bytes = filestore::fetch_concat_exact(&blobstore, ctx, submodule_file_content_id, 20)
        .await
        .with_context(
            || "Failed to fetch content of file containing the submodule's git commit hash",
        )?;

    let git_submodule_hash = RichGitSha1::from_bytes(&bytes, ObjectKind::Commit.as_str(), 0)?;
    let git_submodule_sha1 = git_submodule_hash.sha1();

    anyhow::Ok(git_submodule_sha1)
}

/// Get the git hash from a submodule file, which represents the commit from the
/// given submodule that the source repo depends on at that revision.
pub(crate) async fn git_hash_from_submodule_metadata_file<'a>(
    ctx: &'a CoreContext,
    large_repo: &'a InMemoryRepo,
    submodule_file_content_id: ContentId,
) -> Result<GitSha1> {
    let bytes = filestore::fetch_concat_exact(large_repo.repo_blobstore(), ctx, submodule_file_content_id, 40)
      .await
      .with_context(|| {
          format!(
              "Failed to fetch content from content id {} file containing the submodule's git commit hash",
              &submodule_file_content_id
          )
      })?;

    let git_hash_string = std::str::from_utf8(bytes.as_ref())?;
    let git_sha1 = GitSha1::from_str(git_hash_string)?;

    anyhow::Ok(git_sha1)
}

pub(crate) fn get_submodule_repo<'a, 'b, R: Repo>(
    sm_path: &'a SubmodulePath,
    submodule_deps: &'b HashMap<NonRootMPath, R>,
) -> Result<&'b R> {
    submodule_deps
        .get(&sm_path.0)
        .ok_or_else(|| anyhow!("Mononoke repo from submodule {} not available", sm_path.0))
}

/// Returns true if the given path is a git submodule.
pub(crate) async fn is_path_git_submodule(
    ctx: &CoreContext,
    repo: &impl Repo,
    changeset: ChangesetId,
    path: &NonRootMPath,
) -> Result<bool, Error> {
    Ok(get_submodule_file_content_id(ctx, repo, changeset, path)
        .await?
        .is_some())
}

pub(crate) fn x_repo_submodule_metadata_file_basename<S: std::fmt::Display>(
    submodule_basename: &S,
    x_repo_submodule_metadata_file_prefix: &str,
) -> Result<MPathElement> {
    MPathElement::new(
        format!(".{x_repo_submodule_metadata_file_prefix}-{submodule_basename}")
            .to_string()
            .into_bytes(),
    )
}
/// Builds the full path of the x-repo submodule metadata file for a given
/// submodule.
pub(crate) fn get_x_repo_submodule_metadata_file_path(
    submodule_file_path: &SubmodulePath,
    // Prefix used to generate the metadata file basename. Obtained from
    // the small repo sync config.
    x_repo_submodule_metadata_file_prefix: &str,
) -> Result<NonRootMPath> {
    let (mb_sm_parent_dir, sm_basename) = submodule_file_path.0.split_dirname();

    let x_repo_sm_metadata_file = x_repo_submodule_metadata_file_basename(
        &sm_basename,
        x_repo_submodule_metadata_file_prefix,
    )?;

    let x_repo_sm_metadata_path = match mb_sm_parent_dir {
        Some(sm_parent_dir) => sm_parent_dir.join(&x_repo_sm_metadata_file),
        None => x_repo_sm_metadata_file.into(),
    };
    Ok(x_repo_sm_metadata_path)
}

// Returns the differences between a submodule commit and its parents.
pub(crate) async fn submodule_diff(
    ctx: &CoreContext,
    sm_repo: &impl Repo,
    cs_id: ChangesetId,
    parents: Vec<ChangesetId>,
) -> Result<impl Stream<Item = Result<BonsaiDiffFileChange<(ContentId, u64)>>>> {
    let fsnode_id = sm_repo
        .repo_derived_data()
        .derive::<RootFsnodeId>(ctx, cs_id)
        .await
        .with_context(|| format!("Failed to get fsnode id form changeset id {}", cs_id))?
        .into_fsnode_id();

    let parent_fsnode_ids = stream::iter(parents)
        .then(|parent_cs_id| async move {
            anyhow::Ok(
                sm_repo
                    .repo_derived_data()
                    .derive::<RootFsnodeId>(ctx, parent_cs_id)
                    .await
                    .with_context(|| {
                        format!(
                            "Failed to get parent's fsnode id from its changeset id: {}",
                            parent_cs_id
                        )
                    })?
                    .into_fsnode_id(),
            )
        })
        .try_collect::<HashSet<_>>()
        .await?;

    Ok(bonsai_diff(
        ctx.clone(),
        sm_repo.repo_blobstore_arc().clone(),
        fsnode_id,
        parent_fsnode_ids,
    ))
}

/// Returns the content id of the given path if it is a submodule file.
pub(crate) async fn get_submodule_file_content_id(
    ctx: &CoreContext,
    repo: &impl Repo,
    cs_id: ChangesetId,
    path: &NonRootMPath,
) -> Result<Option<ContentId>> {
    content_id_of_file_with_type(ctx, repo, cs_id, path, FileType::GitSubmodule).await
}

/// Returns the content id of a file at a given path if it was os a specific
/// file type.
pub(crate) async fn content_id_of_file_with_type<R>(
    ctx: &CoreContext,
    repo: &R,
    cs_id: ChangesetId,
    path: &NonRootMPath,
    expected_file_type: FileType,
) -> Result<Option<ContentId>>
where
    R: RepoDerivedDataRef + RepoBlobstoreArc,
{
    let fsnode_id = repo
        .repo_derived_data()
        .derive::<RootFsnodeId>(ctx, cs_id)
        .await?
        .into_fsnode_id();

    let entry = fsnode_id
        .find_entry(ctx.clone(), repo.repo_blobstore_arc(), path.clone().into())
        .await?;

    match entry {
        Some(Entry::Leaf(file)) if *file.file_type() == expected_file_type => {
            Ok(Some(file.content_id().clone()))
        }
        _ => Ok(None),
    }
}

pub(crate) async fn list_all_paths(
    ctx: &CoreContext,
    repo: &impl Repo,
    changeset: ChangesetId,
) -> Result<impl Stream<Item = Result<NonRootMPath>>> {
    let fsnode_id = repo
        .repo_derived_data()
        .derive::<RootFsnodeId>(ctx, changeset)
        .await?
        .into_fsnode_id();
    Ok(fsnode_id
        .list_leaf_entries(ctx.clone(), repo.repo_blobstore_arc())
        .map_ok(|(path, _)| path))
}

pub(crate) async fn list_non_submodule_files_under<R>(
    ctx: &CoreContext,
    repo: &R,
    cs_id: ChangesetId,
    submodule_path: SubmodulePath,
) -> Result<impl Stream<Item = Result<NonRootMPath>>>
where
    R: RepoDerivedDataRef + RepoBlobstoreArc,
{
    let fsnode_id = repo
        .repo_derived_data()
        .derive::<RootFsnodeId>(ctx, cs_id)
        .await?
        .into_fsnode_id();

    Ok(fsnode_id
        .list_leaf_entries_under(
            ctx.clone(),
            repo.repo_blobstore_arc(),
            vec![submodule_path.0],
        )
        .try_filter_map(|(path, fsnode_file)| {
            future::ready(Ok(
                (*fsnode_file.file_type() != FileType::GitSubmodule).then_some(path)
            ))
        }))
}

/// Gets the root directory's fsnode id from a submodule commit provided as
/// as a git hash. This is used for working copy validation of submodule
/// expansion.
pub(crate) async fn root_fsnode_id_from_submodule_git_commit(
    ctx: &CoreContext,
    repo: &impl Repo,
    git_hash: GitSha1,
) -> Result<FsnodeId> {
    let cs_id: ChangesetId = repo
        .bonsai_git_mapping()
        .get_bonsai_from_git_sha1(ctx, git_hash)
        .await?
        .ok_or_else(|| {
            anyhow!(
                "Failed to get changeset id from git submodule commit hash {} in repo {}",
                &git_hash,
                &repo.repo_identity().name()
            )
        })?;
    let submodule_root_fsnode_id: RootFsnodeId = repo
        .repo_derived_data()
        .derive::<RootFsnodeId>(ctx, cs_id)
        .await?;

    Ok(submodule_root_fsnode_id.into_fsnode_id())
}

pub async fn run_and_log_stats_to_scuba<R, S>(
    ctx: &CoreContext,
    log_tag: &str,
    msg: S,
    fut: impl Future<Output = R>,
) -> R
where
    S: Into<Option<String>>,
{
    let (stats, result) = fut.timed().await;
    let mut scuba = ctx.scuba().clone();
    scuba.add_future_stats(&stats);
    scuba.log_with_msg(log_tag, msg);
    result
}

/// Build a new submodule dependency map to expand/validate recursive submodules
/// under a given submodule.
/// It removes the path of the given submodule from all the entries that are
/// under it and ignores the ones that aren't.
pub(crate) fn build_recursive_submodule_deps<R: Repo>(
    submodule_deps: &HashMap<NonRootMPath, R>,
    submodule_path: &NonRootMPath,
) -> HashMap<NonRootMPath, R> {
    let rec_small_repo_deps: HashMap<NonRootMPath, R> = submodule_deps
        .iter()
        .filter_map(|(p, repo)| {
            p.remove_prefix_component(submodule_path)
                .map(|relative_p| (relative_p, repo.clone()))
        })
        .collect();

    rec_small_repo_deps
}
