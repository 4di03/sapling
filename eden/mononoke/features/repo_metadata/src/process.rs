/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use anyhow::anyhow;
use anyhow::Result;
use bookmarks::BookmarkKey;
use context::CoreContext;
use futures::Stream;
use mononoke_types::ChangesetId;

use crate::types::MetadataItem;
use crate::Repo;

pub(crate) async fn process_bookmark<'a>(
    ctx: &'a CoreContext,
    repo: &'a impl Repo,
    bookmark: &BookmarkKey,
) -> Result<impl Stream<Item = Result<MetadataItem>> + 'a> {
    let cs_id = repo
        .bookmarks()
        .get(ctx.clone(), bookmark)
        .await?
        .ok_or_else(|| {
            anyhow!(
                "Bookmark {} not found for repo {}",
                bookmark,
                repo.repo_identity().name()
            )
        })?;

    process_changeset(ctx, repo, cs_id).await
}

async fn process_changeset<'a>(
    _ctx: &'a CoreContext,
    _repo: &'a impl Repo,
    _cs_id: ChangesetId,
) -> Result<impl Stream<Item = Result<MetadataItem>> + 'a> {
    Ok(futures::stream::once(async { Ok(MetadataItem::Unknown) }))
}
