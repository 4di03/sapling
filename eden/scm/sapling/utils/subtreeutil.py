# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

import json
from dataclasses import dataclass
from enum import Enum
from operator import itemgetter

from .. import error, node, pathutil, util
from ..i18n import _

# this config is for testing purpose only
CFG_ALLOW_ANY_SOURCE_COMMIT = "allow-any-source-commit"

# todo: remove the 'test_' prefix when this feature is stable
SUBTREE_BRANCH_KEY = "test_subtree_copy"
SUBTREE_MERGE_KEY = "test_subtree_merge"

# keys that are used for subtree operations, this list should
# include the keys for both O(n) copy and O(1) copy
SUBTREE_OPERATION_KEYS = [
    SUBTREE_BRANCH_KEY,
    SUBTREE_MERGE_KEY,
]


class BranchType(Enum):
    DEEP_COPY = 1  # O(n) subtree copy
    SHALLOW_COPY = 2  # O(1) subtree copy

    def to_key(self):
        # the `key` is used in subtree metadata
        if self == BranchType.DEEP_COPY:
            return "deepcopies"
        elif self == BranchType.SHALLOW_COPY:
            return "copies"
        else:
            # unreachable
            raise error.ProgrammingError("unknown branch type")


@dataclass
class SubtreeBranch:
    version: int
    branch_type: BranchType
    from_commit: str
    from_path: str
    to_path: str

    def to_dict(self):
        return {
            "from_commit": self.from_commit,
            "from_path": self.from_path,
            "to_path": self.to_path,
        }


@dataclass
class SubtreeMerge:
    version: int
    from_commit: str
    from_path: str
    to_path: str

    def to_dict(self):
        return {
            "from_commit": self.from_commit,
            "from_path": self.from_path,
            "to_path": self.to_path,
        }


def gen_branch_info(repo, from_commit, from_paths, to_paths):
    if not is_source_commit_allowed(repo.ui, repo[from_commit]):
        return {}

    # sort by to_path
    path_mapping = sorted(zip(from_paths, to_paths), key=itemgetter(1))
    value = {
        "v": 1,
        "branches": [
            {
                "from_commit": from_commit,
                "from_path": from_path,
                "to_path": to_path,
            }
            for from_path, to_path in path_mapping
        ],
    }
    # compact JSON representation
    str_val = json.dumps(value, separators=(",", ":"), sort_keys=True)
    return {SUBTREE_BRANCH_KEY: str_val}


def gen_merge_info(repo, subtree_merges):
    subtree_merges = [
        m for m in subtree_merges if is_source_commit_allowed(repo.ui, repo[m[0]])
    ]
    if not subtree_merges:
        return {}

    # sort by to_path
    subtree_merges = sorted(subtree_merges, key=itemgetter(2))
    value = {
        "v": 1,
        "merges": [
            {
                "from_commit": node.hex(from_node),
                "from_path": from_path,
                "to_path": to_path,
            }
            for from_node, from_path, to_path in subtree_merges
        ],
    }
    # compact JSON representation
    str_val = json.dumps(value, separators=(",", ":"), sort_keys=True)
    return {SUBTREE_MERGE_KEY: str_val}


def get_subtree_metadata(extra):
    """Get the subtree metadata from commit's extra."""
    return {k: v for k, v in extra.items() if k in SUBTREE_OPERATION_KEYS}


def get_branch_info(repo, node):
    return _get_subtree_metadata(repo, node, SUBTREE_BRANCH_KEY)


def get_merge_info(repo, node):
    return _get_subtree_metadata(repo, node, SUBTREE_MERGE_KEY)


def _get_subtree_metadata(repo, node, key):
    extra = repo[node].extra()
    try:
        val_str = extra[key]
    except KeyError:
        return None
    try:
        return json.loads(val_str)
    except json.JSONDecodeError:
        raise error.Abort(f"invalid {key} metadata: {val_str}")


def validate_path_exist(ui, ctx, paths, abort_on_missing=False):
    """Validate that the given path exists in the given context."""
    for p in paths:
        if not (p in ctx or ctx.hasdir(p)):
            msg = _("path '%s' does not exist in commit %s") % (p, ctx)
            if abort_on_missing:
                raise error.Abort(msg)
            else:
                ui.status(msg + "\n")


def validate_path_size(from_paths, to_paths, abort_on_empty=False):
    if len(from_paths) != len(to_paths):
        raise error.Abort(_("must provide same number of --from-path and --to-path"))

    if abort_on_empty and not from_paths:
        raise error.Abort(_("must provide --from-path and --to-path"))


def validate_path_overlap(from_paths, to_paths):
    # Disallow overlapping --to-path to keep things simple.
    to_dirs = util.dirs(to_paths)
    seen = set()
    for p in to_paths:
        if p in to_dirs or p in seen:
            raise error.Abort(_("overlapping --to-path entries"))
        seen.add(p)

    from_dirs = util.dirs(from_paths)
    for from_path, to_path in zip(from_paths, to_paths):
        if from_path in to_dirs or to_path in from_dirs:
            raise error.Abort(
                _("overlapping --from-path '%s' and --to-path '%s'")
                % (from_path, to_path)
            )


def find_enclosing_dest(target_path, paths):
    """Find the path that contains the target path.

    >>> is_in_subtree_copy_dest("a/b/c", ["a/b"])
    'a/b'
    >>> is_in_subtree_copy_dest("a/b/c", ["a/b/c"])
    'a/b/c'
    >>> is_in_subtree_copy_dest("a/b/c", ["a/b", "e/f"])
    'a/b'
    >>> is_in_subtree_copy_dest("a/b/c", ["a/b/c/d", "e/f"])
    """
    target_dir = pathutil.dirname(target_path)
    for path in paths:
        if target_dir.startswith(path) or path == target_path:
            return path
    return None


def validate_source_commit(ui, source_ctx, subcmd_name):
    if is_source_commit_allowed(ui, source_ctx):
        return

    if educationpage := ui.config("subtree", "education-page"):
        hint = _("see subtree %s at %s for the impacts on subtree merge and log") % (
            subcmd_name,
            educationpage,
        )
    else:
        hint = _("see '@prog@ help subtree' for the impacts on subtree merge and log")
    prompt_msg = _(
        "subtree %s from a non-public commit is not recommended. However, you can\n"
        "still proceed and use subtree copy and merge for common cases.\n"
        "(hint: %s)\n"
        "Continue with subtree %s (y/n)? $$ &Yes $$ &No"
    ) % (subcmd_name, hint, subcmd_name)
    if ui.promptchoice(prompt_msg, default=1) != 0:
        raise error.Abort(
            f"subtree {subcmd_name} from a non-public commit is not allowed"
        )


def is_source_commit_allowed(ui, source_ctx) -> bool:
    # Currently, we only allow public commits as source commits
    # later, we will allow ancestor-draft commits as well.
    if ui.configbool("subtree", CFG_ALLOW_ANY_SOURCE_COMMIT):
        return True
    if source_ctx.ispublic():
        return True
    return False
