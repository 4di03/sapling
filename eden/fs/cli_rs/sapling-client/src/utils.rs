/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::collections::HashMap;
use std::env;
use std::ffi::OsString;
use std::fs::read_to_string;
use std::path::Path;

pub(crate) fn get_sapling_executable_path() -> String {
    let path = env::var("EDEN_HG_BINARY").unwrap_or_else(|_| String::new());
    if path.is_empty() {
        "hg".to_string() // `sl` is not always available, so use `hg`
    } else {
        path
    }
}

pub(crate) fn get_sapling_options() -> HashMap<OsString, OsString> {
    let mut options = HashMap::<OsString, OsString>::new();
    // Ensure that the hgrc doesn't mess with the behavior of the commands that we're running.
    options.insert("HGPLAIN".to_string().into(), "1".to_string().into());
    // Ensure that we do not log profiling data for the commands we are
    // running. This is to avoid a significant increase in the rate of logging.
    options.insert("NOSCMLOG".to_string().into(), "1".to_string().into());
    // chg can elect to kill all children if an error occurs in any child.
    // This can cause commands we spawn to fail transiently.  While we'd
    // love to have the lowest latency, the transient failure causes problems
    // with our ability to deliver notifications to our clients in a timely
    // manner, so we disable the use of chg for the sapling processes
    // that we spawn.
    options.insert("CHGDISABLE".to_string().into(), "1".to_string().into());
    options
}

pub fn is_fbsource_checkout(mount_point: &Path) -> bool {
    let project_id_path = mount_point.join(".projectid");
    let project_id = read_to_string(project_id_path).ok();
    match project_id {
        Some(project_id) => project_id.trim() == "fbsource",
        None => false,
    }
}

#[cfg(test)]
mod tests {
    use edenfs_client::utils::get_mount_point;

    use crate::utils::*;

    #[test]
    pub fn test_is_fbsource_checkout() -> anyhow::Result<()> {
        let mount_point = get_mount_point(&None)?;
        assert!(is_fbsource_checkout(&mount_point));
        Ok(())
    }
}
