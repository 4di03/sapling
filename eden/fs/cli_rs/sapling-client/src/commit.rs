/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use tokio::process::Command;

use crate::get_sapling_executable_path;
use crate::get_sapling_options;

pub async fn get_current_commit_id() -> anyhow::Result<String> {
    let output = Command::new(get_sapling_executable_path())
        .envs(get_sapling_options())
        .args(["whereami", "--traceback"])
        .output()
        .await?;
    Ok(String::from_utf8(output.stdout)?)
}

pub async fn get_commit_timestamp(commit_id: &str) -> anyhow::Result<u64> {
    let output = Command::new(get_sapling_executable_path())
        .envs(get_sapling_options())
        .args(["log", "--traceback", "-T", "{date}", "-r", commit_id])
        .output()
        .await?;
    let date = String::from_utf8(output.stdout)?;
    let date = date.parse::<f64>().map_err(anyhow::Error::msg)?;
    // NOTE: Sapling returns fractional seconds, but we only want whole seconds.
    //       Truncate to the nearest second.
    Ok(date as u64)
}

pub async fn is_commit_in_repo(commit_id: &str) -> anyhow::Result<bool> {
    let output = Command::new(get_sapling_executable_path())
        .envs(get_sapling_options())
        .args([
            "log",
            "--traceback",
            "-r",
            commit_id,
            // Disable fbsource <-> www sync during lookup
            "--config",
            "megarepo.transparent-lookup=",
        ])
        .output()
        .await?;
    Ok(output.status.success())
}

#[cfg(test)]
mod tests {
    use edenfs_client::utils::get_mount_point;

    use crate::commit::*;
    use crate::*;

    const FBSOURCE_COMMIT_ID: &str = "5496dd87e5fe7430a1a399530cc339a479097524";
    const WWW_COMMIT_ID: &str = "1061662d6db2072dd30308d1626a45ac11db3467";

    #[tokio::test]
    pub async fn test_current_commit_id() -> anyhow::Result<()> {
        let commit_id = get_current_commit_id().await?;
        assert!(!commit_id.is_empty());
        Ok(())
    }

    #[tokio::test]
    pub async fn test_is_commit_in_repo() -> anyhow::Result<()> {
        let mount_point = get_mount_point(&None)?;
        let commit_id = get_current_commit_id().await?;
        assert!(is_commit_in_repo(&commit_id).await?);
        assert_eq!(
            is_commit_in_repo(FBSOURCE_COMMIT_ID).await?,
            is_fbsource_checkout(&mount_point)
        );
        assert_eq!(
            is_commit_in_repo(WWW_COMMIT_ID).await?,
            !is_fbsource_checkout(&mount_point)
        );

        Ok(())
    }

    #[tokio::test]
    pub async fn test_get_commit_timestamp() -> anyhow::Result<()> {
        // sl log of commit in fbsource:
        //   changeset:   5496dd87e5fe7430a1a399530cc339a479097524  D68746950
        //   user:        John Elliott <jdelliot@fb.com>
        //   date:        Tue, 28 Jan 2025 10:35:17 -0800
        //   summary:     [meerkat] Improve saved state support
        // timestamp should be 1738089317.028800, but we truncate to the nearest second
        let timestamp = get_commit_timestamp(FBSOURCE_COMMIT_ID).await?;
        assert_eq!(timestamp, 1738089317);
        Ok(())
    }
}
