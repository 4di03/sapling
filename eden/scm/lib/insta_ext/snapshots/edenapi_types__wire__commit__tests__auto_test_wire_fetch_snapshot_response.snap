---
source: commit.rs
expression: "WireFetchSnapshotResponse :: arbitrary(& mut g)"
---
{"2":[["aiffeu/uatbfkh/jkleav",{"5":[{"1":{"1":{"4":[96,254,202,4,46,38,85,25,36,58,55,101,239,1,100,81,21,29,1,4]},"3":6325243916544195815}},"3",["gdqp/nhxnr/ixfpat/mw",18446744073709551615]]}]],"4":5294005526868283990,"5":-1808809550,"6":14415315640882124868,"7":[""]}
